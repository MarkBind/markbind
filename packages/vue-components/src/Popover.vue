<template>
  <span
    :class="trigger === 'click' ? 'trigger-click' : 'trigger'"
    data-mb-component-type="popover"
    tabindex="0"
  >
    <portal v-if="targetEl.id" :to="'popover:' + targetEl.id">
      <h3 v-if="hasHeader" class="popover-header">
        <slot name="header"></slot>
      </h3>
      <div class="popover-body">
        <slot name="content"></slot>
      </div>
    </portal><!-- do not delete this comment, it is for the stray space issue (#2419)
 --><v-popover
      v-if="isMounted"
      :auto-hide="!isInput"
      :triggers="triggers"
      :popper-triggers="triggers"
      :hide-triggers="triggers"
      :placement="placement"
      :delay="0"
      shift-cross-axis
    >
      <!-- floating-vue triggers must be elements that receive mouse events, hence an empty @click -->
      <span v-if="!isInput" @click.stop>
        <slot></slot>
      </span>
      <!-- However, input elements are handled separately as they will lose focus when wrapped in a span -->
      <slot v-else></slot>
      <template #popper>
        <div class="popover-container">
          <h3 v-if="hasHeader" class="popover-header">
            <slot name="header"></slot>
          </h3>
          <div class="popover-body">
            <slot name="content"></slot>
          </div>
        </div>
      </template>
    </v-popover>
  </span>
</template>

<script>
// eslint-disable-next-line import/no-extraneous-dependencies
import { Portal } from 'portal-vue';

export default {
  name: 'Popover',
  components: {
    Portal,
  },
  props: {
    trigger: {
      type: String,
      default: 'hover focus',
    },
    placement: {
      type: String,
      default: 'top',
    },
  },
  data() {
    return {
      targetEl: {},
      isMounted: false,
    };
  },
  computed: {
    triggers() {
      return this.trigger.split(' ');
    },
    hasHeader() {
      return !!this.$slots.header;
    },
  },
  mounted() {
    this.targetEl = this.$el;
    // <input> tags need to be handled separately as they need to retain focus on inputs
    this.isInput = this.$slots.default && this.$slots.default.some(node => node.tag === 'input');
    this.isMounted = true;
  },
};
</script>

<style>
    .popover-container {
        overflow: auto;
        max-height: 50vh;
        max-width: 276px;  /* following bootstrap */
    }

    .popover-body {
        font-size: 0.875rem;  /* following bootstrap */
    }

    .v-popper {
        display: inline;
    }
</style>
