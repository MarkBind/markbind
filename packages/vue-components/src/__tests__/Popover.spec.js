import { mount } from '@vue/test-utils';
import Popover from '../Popover.vue';

const DEFAULT_POPOVER_TRIGGER = 'click to show popover';
const DEFAULT_POPOVER_HEADER_ATTRIBUTE = 'test popover header attribute';
const DEFAULT_POPOVER_HEADER_SLOT = 'test popover header slot';
const DEFAULT_POPOVER_CONTENT_ATTRIBUTE = 'Popover content slot: Lorem ipsum dolor sit amet, '
+ 'consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.';
const DEFAULT_POPOVER_CONTENT_SLOT = 'Popover content slot: Lorem ipsum dolor sit amet, '
+ 'consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.';

describe('Popover', () => {
  test('should not be shown until triggered', async () => {
    const wrapper = mount(Popover, {
      propsData: {
        trigger: 'click',
        content: DEFAULT_POPOVER_CONTENT_ATTRIBUTE,
      },
      slots: {
        default: DEFAULT_POPOVER_TRIGGER,
      },
      stubs: ['v-popover'],
    });

    // don't click on trigger
    expect(wrapper.element).toMatchSnapshot();
  });

  test('should not show header when no header is given', async () => {
    const wrapper = mount(Popover, {
      propsData: {
        trigger: 'click',
        content: DEFAULT_POPOVER_CONTENT_ATTRIBUTE,
      },
      slots: {
        default: DEFAULT_POPOVER_TRIGGER,
      },
      stubs: ['v-popover'],
    });

    // click on trigger
    await wrapper.find('span.trigger-click').trigger('click');

    expect(wrapper.element).toMatchSnapshot();
  });

  test('should show header and content', async () => {
    const wrapper = mount(Popover, {
      propsData: {
        trigger: 'click',
        content: DEFAULT_POPOVER_CONTENT_ATTRIBUTE,
        header: DEFAULT_POPOVER_HEADER_ATTRIBUTE,
      },
      slots: {
        default: DEFAULT_POPOVER_TRIGGER,
      },
      stubs: ['v-popover'],
    });

    // click on trigger
    await wrapper.find('span.trigger-click').trigger('click');

    expect(wrapper.element).toMatchSnapshot();
  });

  test('should show header and content slots correctly', async () => {
    const wrapper = mount(Popover, {
      propsData: {
        trigger: 'click',
      },
      slots: {
        default: DEFAULT_POPOVER_TRIGGER,
        header: DEFAULT_POPOVER_HEADER_SLOT,
        content: DEFAULT_POPOVER_CONTENT_SLOT,
      },
      stubs: ['v-popover'],
    });

    // click on trigger
    await wrapper.find('span.trigger-click').trigger('click');

    expect(wrapper.element).toMatchSnapshot();
  });

  test('should respect the placement attribute', async () => {
    const wrapper = mount(Popover, {
      propsData: {
        trigger: 'click',
        content: DEFAULT_POPOVER_CONTENT_ATTRIBUTE,
        header: DEFAULT_POPOVER_HEADER_ATTRIBUTE,
        placement: 'right',
      },
      slots: {
        default: DEFAULT_POPOVER_TRIGGER,
      },
      stubs: ['v-popover'],
    });

    // click on trigger
    await wrapper.find('span.trigger-click').trigger('click');

    expect(wrapper.element).toMatchSnapshot();
  });
});
