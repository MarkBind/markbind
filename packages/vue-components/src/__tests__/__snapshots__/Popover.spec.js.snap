// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should not be shown until triggered 1`] = `
<span
  class="trigger-click"
  content="Popover content slot: Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
  data-mb-component-type="popover"
  tabindex="0"
>
  <!---->
  <!---->
</span>
`;

exports[`Popover should not show header when no header is given 1`] = `
<span
  class="trigger-click"
  content="Popover content slot: Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
  data-mb-component-type="popover"
  tabindex="0"
>
  <!---->
  <v-popover-stub
    auto-hide="true"
    delay="0"
    hide-triggers="click"
    placement="top"
    popper-triggers="click"
    shift-cross-axis=""
    triggers="click"
  >
    <span>
      click to show popover
    </span>
  </v-popover-stub>
</span>
`;

exports[`Popover should respect the placement attribute 1`] = `
<span
  class="trigger-click"
  content="Popover content slot: Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
  data-mb-component-type="popover"
  header="test popover header attribute"
  tabindex="0"
>
  <!---->
  <v-popover-stub
    auto-hide="true"
    delay="0"
    hide-triggers="click"
    placement="right"
    popper-triggers="click"
    shift-cross-axis=""
    triggers="click"
  >
    <span>
      click to show popover
    </span>
  </v-popover-stub>
</span>
`;

exports[`Popover should show header and content 1`] = `
<span
  class="trigger-click"
  content="Popover content slot: Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
  data-mb-component-type="popover"
  header="test popover header attribute"
  tabindex="0"
>
  <!---->
  <v-popover-stub
    auto-hide="true"
    delay="0"
    hide-triggers="click"
    placement="top"
    popper-triggers="click"
    shift-cross-axis=""
    triggers="click"
  >
    <span>
      click to show popover
    </span>
  </v-popover-stub>
</span>
`;

exports[`Popover should show header and content slots correctly 1`] = `
<span
  class="trigger-click"
  data-mb-component-type="popover"
  tabindex="0"
>
  <!---->
  <v-popover-stub
    auto-hide="true"
    delay="0"
    hide-triggers="click"
    placement="top"
    popper-triggers="click"
    shift-cross-axis=""
    triggers="click"
  >
    <span>
      click to show popover
    </span>
  </v-popover-stub>
</span>
`;
